packages/yew-router/src/hidden.rs

```diff
@@ -0,0 +1,407 @@
+use std::any::{Any, TypeId};
+use std::borrow::Cow;
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::convert::{TryFrom, TryInto};
+use std::fmt::Display;
+use std::mem;
+use std::str::FromStr;
+
+use percent_encoding::{percent_decode_str, percent_encode, NON_ALPHANUMERIC};
+use regex::RegexSet;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use wasm_bindgen::JsValue;
+
+use crate::history::Route;
+pub use crate::Routable;
+
+/// This trait is implemented by the derive macro. There is a blanket implementation of
+/// `Routeable` for all `DerivedRoutable`.
+pub trait DerivedRoutable: PartialEq + Any + Default + Clone {
+    const VARIANTS: &'static [RouteMuncher<Self>];
+
+    fn to_args(&self) -> (usize, Args);
+}
+
+/// Lazily builds and caches a `RegexSet` keyed on the TypeId of the `DerivedRoutable`.
+fn cached_regex_set<T: DerivedRoutable>() -> &'static RegexSet {
+    thread_local! {
+        static REGEX_CACHE: RefCell<HashMap<TypeId, &'static RegexSet>> = RefCell::new(HashMap::new());
+    }
+
+    // In the fast path, we simply lookup the TypeId in the map.
+    REGEX_CACHE.with(|cache| {
+        *cache
+            .borrow_mut()
+            .entry(TypeId::of::<T>())
+            .or_insert_with(|| {
+                // In the slow path we build a `RegexSet` from all the individual variant regexes. Leaking
+                // this `RegexSet` is fine, because the total number of distinct `DerivedRoutable` types is
+                // fixed at compile time.
+                Box::leak(Box::new(
+                    RegexSet::new(T::VARIANTS.iter().map(|v| v.regex))
+                        .expect("Invalid routing regex"),
+                ))
+            })
+    })
+}
+
+/// Blanket implementation of `Routable` for `DerivedRoutable`.
+impl<T: DerivedRoutable> Routable for T {
+    fn from_route(route: &Route) -> Result<Self, Self>
+    where
+        Self: Sized,
+    {
+        // Use a `RegexSet` to rapidly narrow the search to a small number of
+        // candidate routes.
+        let regex_set = cached_regex_set::<T>();
+        for candidate in regex_set.matches(&route.path) {
+            // Try each route in turn. `RegexSet` guarantees iteration order to
+            // match the original array.
+            let variant = &T::VARIANTS[candidate];
+            if let Some(res) = variant.from_route(route) {
+                return Ok(res);
+            }
+        }
+        Err(T::default())
+    }
+
+    fn to_route(&self) -> Route {
+        let (index, args) = self.to_args();
+        let variant = &T::VARIANTS[index];
+        variant.reconstitute(args)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum Arg {
+    Empty,
+    String(String),
+    Opaque(JsValue),
```

Comment:
Why is a `JsValue` in the exposed API?

Reply:
This is a remnant from when "state" was supported. It's just the API exposed to the macro, it's hidden from the public API. I'll remove it for now, but keep in mind this potential for expansion is why `Arg` is an enum rather than an `Option<String>`.